# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [48]:
def najdi_prvocisla(n, prvocisla = [2]):
    i = prvocisla[-1]
    while len(prvocisla) < n:
        i += 1
        for prvocislo in prvocisla:
            if i % prvocislo == 0:
                break
        else:
            prvocisla.append(i)
    return prvocisla[:n]
    

print(najdi_prvocisla(12))
print(najdi_prvocisla(50))
print(najdi_prvocisla(4))

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229]
[2, 3, 5, 7]


In [49]:
najdi_prvocisla.__defaults__

([2,
  3,
  5,
  7,
  11,
  13,
  17,
  19,
  23,
  29,
  31,
  37,
  41,
  43,
  47,
  53,
  59,
  61,
  67,
  71,
  73,
  79,
  83,
  89,
  97,
  101,
  103,
  107,
  109,
  113,
  127,
  131,
  137,
  139,
  149,
  151,
  157,
  163,
  167,
  173,
  179,
  181,
  191,
  193,
  197,
  199,
  211,
  223,
  227,
  229],)


# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [31]:
%%writefile muj_modul.py
def odmocnina(a, n): 
    x = float(a)
    for i in range(n):
        x = (a/x + x)/2
        # print(x)
    return x

def odhad_pi(n):
    a = 1
    b = 1
    v = odmocnina((a**2 - (b/2)**2), 100)
    
    for i in range(n):
        b = odmocnina(((b/2)**2 + ((1 - v)**2)), 100)
        v = odmocnina((a**2 - (b/2)**2), 100)
        # S = 6 * (2 ** (i+1)) * b * v/2
        # print(S)

    S = 6 * (2 ** n) * b * v/2
    return S

def aproximace_pi(n):
    a = 1/(2 ** 4)
    suma = a/3
    for i in range(n): 
        a = a * (2 * (i + 2) - 3) / ((2 ** 3) * (i + 2))
        suma = suma + a / (2 * (i+2) + 1)
    moje_pi = 12 * (-odmocnina(3,1000)/8 + 1/2 - suma)
    return moje_pi

Overwriting muj_modul.py


In [50]:
import muj_modul
# from muj_modul import aproximace_pi, odhad_pi, odmocnina

In [51]:
dir(muj_modul)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'aproximace_pi',
 'odhad_pi',
 'odmocnina']

In [32]:
%ll

celkem 844
-rw-rw-r-- 1 student    118 dub  3 19:55 modul.py
-rw-rw-r-- 1 student    685 dub  3 20:11 muj_modul.py
-rw-rw-r-- 1 student 218489 dub  3 10:41 01_ukoly.ipynb
-rw-rw-r-- 1 student  12596 dub  3 18:39 02_ukoly.ipynb
-rw-rw-r-- 1 student 477534 bře 24 13:16 05_scipy.ipynb
-rw-rw-r-- 1 student 123381 bře 24 13:50 05_sympy.ipynb
-rw-rw-r-- 1 student  10052 bře 24 14:02 05_ukoly.ipynb


In [33]:
%run muj_modul.py

In [34]:
%who

aproximace_pi	 i	 najdi_prvocisla	 odhad_pi	 odmocnina	 prvocisla	 


In [37]:
%time odhad_pi(100)
%time aproximace_pi(100)

CPU times: user 13.3 ms, sys: 0 ns, total: 13.3 ms
Wall time: 47.5 ms
CPU times: user 732 µs, sys: 0 ns, total: 732 µs
Wall time: 737 µs


3.1415926535897936

In [40]:
%timeit odhad_pi(100)

8.26 ms ± 852 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [42]:
%%time 
odhad_pi(100)
odhad_pi(10)

CPU times: user 14.3 ms, sys: 0 ns, total: 14.3 ms
Wall time: 24 ms


3.141592105999271

In [43]:
%%timeit
odhad_pi(100)
odhad_pi(10)

10.5 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [54]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        return Addition(self, other)
    
    def __sub__(self, other):
        return self + cst(-1)*other# TODO (použijte Addition a Multiplication se zápornou konstantu)
    
    def __mul__(self, other):
        return Multiplication(self, other) # TODO (použijte Multiplication)
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"
        
class Addition(BinaryOperator):
    def __str__(self):
        return str(self.levy_argument) + "+" + str(self.pravy_argument) # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
    def diff(self):
        return self.levy_argument.diff() + self.pravy_argument.diff() # TODO
    
class Multiplication(BinaryOperator):
    def __str__(self):
        return "(" + str(self.levy_argument) + ") * (" + str(self.pravy_argument) + ")" # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
    def diff(self):
        return self.levy_argument.diff() * self.pravy_argument + self.pravy_argument.diff() * self.levy_argument# TODO
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return str(self.value) # TODO vypíšeme něco jako value
    def diff(self):
        return cst(0) # TODO
        pass # pass tam být nemusí, smazat
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        return "x" # TODO vypíšeme něco jako x   
    def diff(self):
        return cst(1) # TODO
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        return "(" + str(self.argument) + ")**" + str(self.exponent) # TODO vypíšeme něco jako argument^exponent
    def diff(self):
        return cst(self.exponent) * mocnina(self.argument, self.exponent - 1) * self.argument.diff() # TODO
    
class sin(UnaryOperator):
    def diff(self):
        return cos(self.argument)*self.argument.diff() # TODO

class cos(UnaryOperator):
    def diff(self):
        return cst(-1)*sin(self.argument)*self.argument.diff() # TODO

class exp(UnaryOperator):
    def diff(self):
        return self.argument.diff() * self # TODO
    
class ln(UnaryOperator):
    def diff(self):
        return frc(self.argument)*self.argument.diff() # TODO

class frc(UnaryOperator):
    def __str__(self):
        return "1/(" + str(self.argument) + ")" # TODO vypíšeme něco jako 1/(argument)
    def diff(self):
        return cst(-1) * frc(mocnina(self.argument, 2)) * self.argument.diff() # TODO

In [55]:
x = identity()
f1 = x + cst(1)
print(f1)

x+1


In [57]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

sin(x+cos((x) * (x)))+1
(cos(x+cos((x) * (x)))) * (1+((-1) * (sin((x) * (x)))) * ((1) * (x)+(1) * (x)))+0


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [58]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


(x)**2+(2) * (x)+1+(2) * ((x)**2)
((2) * ((x)**1)) * (1)+(0) * (x)+(1) * (2)+0+(0) * ((x)**2)+(((2) * ((x)**1)) * (1)) * (2)


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [60]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())
print(f3.diff().diff().diff())


(x)**2+(2) * (x)+exp(x)
((2) * ((x)**1)) * (1)+(0) * (x)+(1) * (2)+(1) * (exp(x))
((0) * ((x)**1)+(((1) * ((x)**0)) * (1)) * (0)+(((0) * ((x)**0)+(((0) * ((x)**-1)) * (1)) * (1)) * (1)+(0) * ((1) * ((x)**0))) * (2)+(0) * (((1) * ((x)**0)) * (1))) * (1)+(0) * ((0) * ((x)**1)+(((1) * ((x)**0)) * (1)) * (2))+(0) * ((2) * ((x)**1))+((0) * ((x)**1)+(((1) * ((x)**0)) * (1)) * (2)) * (0)+(0) * (x)+(1) * (0)+(0) * (0)+(0) * (1)+(0) * (2)+(0) * (0)+(0) * (1)+(0) * (0)+(0) * (exp(x))+((1) * (exp(x))) * (0)+((0) * (exp(x))+((1) * (exp(x))) * (1)) * (1)+(0) * ((1) * (exp(x)))


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...